### PR TITLE
Add `ZImagePipeline`

### PR DESCRIPTION
**Pull request by pavel-esir**
## Description
Adds Z-Image Turbo support for text-to-image and image-to-image generation. Z-Image computes normalized transformer timesteps as `(1000 - t) / 1000`, unlike other diffusion pipelines that pass scheduler timesteps directly. Its FlowMatch schedule is also derived from the image sequence length.

The tests are marked xfail this is because `optimum-intel` changes for `z-image` are not merged yet. Ticket for that CVS-178687

CVS-178686 CVS-192190

## Checklist:
- [x] This PR follows [GenAI Contributing guidelines](https://github.com/openvinotoolkit/openvino.genai?tab=contributing-ov-file#contributing). 
- [x]  Tests have been updated or added to cover the new code. 
- [x]  This PR fully addresses the ticket. 
- [ ]  I have made corresponding changes to the documentation. N/A

### CODE REVIEW

File: src/cpp/src/image_generation/models/zimage_transformer_2d_model.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    for (auto&& input : m_model->inputs()) {
+        std::string input_name = input.get_any_name();
+        name_to_shape[input_name] = input.get_partial_shape();
+        if (input_name == "timestep") {
+            name_to_shape[input_name][0] = 1;
+        } else if (input_name == "sample") {
+            name_to_shape[input_name] = {batch_size, m_config.in_channels, height, width};
+        } else if (input_name == "encoder_hidden_states") {
+            name_to_shape[input_name] = {batch_size, 128, name_to_shape[input_name][2]};
+        }
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/models/zimage_transformer_2d_model.cpp]: fixed

File: src/cpp/src/image_generation/models/zimage_transformer_2d_model.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+    std::map<std::string, ov::PartialShape> name_to_shape;
+
+    for (auto&& input : m_model->inputs()) {
+        std::string input_name = input.get_any_name();
+        name_to_shape[input_name] = input.get_partial_shape();
+        if (input_name == "timestep") {
+            name_to_shape[input_name][0] = 1;
+        } else if (input_name == "sample") {
+            name_to_shape[input_name] = {batch_size, m_config.in_channels, height, width};
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/models/zimage_transformer_2d_model.cpp]: applied

File: src/cpp/src/image_generation/zimage_pipeline.hpp [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        custom_generation_config.height = custom_generation_config.height == 0 ? 512 : custom_generation_config.height;
+        custom_generation_config.width = custom_generation_config.width == 0 ? 512 : custom_generation_config.width;
+
+        OPENVINO_ASSERT(custom_generation_config.height == 512 && custom_generation_config.width == 512,
+            "ZImagePipeline only supports fixed resolution of 512x512");
+
+        std::shared_ptr<Generator> generator = custom_generation_config.generator;
+        if (!generator) {
+            OPENVINO_THROW("Generator must be provided for ZImagePipeline");
+        }
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp:281]: fixed

File: src/cpp/src/image_generation/zimage_pipeline.hpp [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+protected:
+    void initialize_generation_config(const std::string& class_name) override {
+        m_generation_config.max_sequence_length = 128;
+        m_generation_config.num_images_per_prompt = 1;
+        m_generation_config.height = 512;
+        m_generation_config.width = 512;
+        m_generation_config.num_inference_steps = 8;
+        m_generation_config.guidance_scale = 7.5f;
+        m_generation_config.negative_prompt = "";
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp]: outdated. this line removed, therefore negative prompt is already initialized with nullopt

File: src/cpp/src/image_generation/models/qwen3_text_encoder.cpp [hunk 5]
```diff
@@ -148,6 +149,39 @@ ov::Tensor Qwen3TextEncoder::infer(const std::string& pos_prompt, const std::str
     m_request.set_tensor("attention_mask", attention_mask);
     m_request.infer();
 
+    for (const ov::Output<const ov::Node>& output : m_request.get_compiled_model().outputs()) {
+        const std::unordered_set<std::string>& output_names = output.get_names();
+        if (output_names.count("last_hidden_state") != 0) {
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/models/qwen3_text_encoder.cpp]: done

File: src/cpp/src/image_generation/zimage_pipeline.hpp [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+        initialize_generation_config("ZImagePipeline");
+    }
+
+    void reshape(const int num_images_per_prompt,
+                 const int height,
+                 const int width,
+                 const float guidance_scale) override {
+        check_image_size(height, width);
+
+        m_text_encoder->reshape(1, m_generation_config.max_sequence_length);
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp]: Outdated. `guidance_scale` is already in use.

File: src/cpp/src/image_generation/models/qwen3_text_encoder.cpp [hunk 7]
```diff
@@ -148,6 +149,39 @@ ov::Tensor Qwen3TextEncoder::infer(const std::string& pos_prompt, const std::str
...
+            ov::Tensor last_hidden_state = m_request.get_tensor("last_hidden_state");
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/models/qwen3_text_encoder.cpp]: The logic below works only when names start with `"hidden_states."` so these additions don't break klean2flux at all. But if i apply your suggestion this will break z-image turbo

File: src/cpp/src/image_generation/models/zimage_transformer_2d_model.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/models/zimage_transformer_2d_model.cpp]: already applied

File: src/cpp/src/image_generation/models/qwen3_text_encoder.cpp [hunk 8]
```diff
@@ -148,6 +149,39 @@ ov::Tensor Qwen3TextEncoder::infer(const std::string& pos_prompt, const std::str
...
+            const ov::Shape hidden_state_shape = last_hidden_state.get_shape();
+            ov::Tensor result(last_hidden_state.get_element_type(), hidden_state_shape);
+            std::memset(result.data(), 0, result.get_byte_size());
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/models/qwen3_text_encoder.cpp]: Applied changes

File: src/cpp/src/image_generation/models/qwen3_text_encoder.cpp [hunk 9]
```diff
@@ -148,6 +149,39 @@ ov::Tensor Qwen3TextEncoder::infer(const std::string& pos_prompt, const std::str
...
+        const std::unordered_set<std::string>& output_names = output.get_names();
+        if (output_names.count("last_hidden_state") != 0 && m_config.hidden_states_layers.empty()) {
+            ov::Tensor last_hidden_state = m_request.get_tensor("last_hidden_state");
+            const ov::Shape hidden_state_shape = last_hidden_state.get_shape();
+            ov::Tensor result(last_hidden_state.get_element_type(), hidden_state_shape);
+            std::memset(result.data(), 0, result.get_byte_size());
+
+            const float* last_hidden_state_data = last_hidden_state.data<const float>();
+            float* result_data = result.data<float>();
+            const size_t hidden_size = hidden_state_shape[2];
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/models/qwen3_text_encoder.cpp:182]: already applied

File: src/cpp/src/image_generation/text2image_pipeline.cpp [hunk 10]
```diff
@@ -33,6 +34,8 @@ Text2ImagePipeline::Text2ImagePipeline(const std::filesystem::path& root_dir) {
         m_impl = std::make_shared<FluxPipeline>(PipelineType::TEXT_2_IMAGE, root_dir);
     } else if (class_name == "Flux2KleinPipeline") {
         m_impl = std::make_shared<Flux2KleinPipeline>(PipelineType::TEXT_2_IMAGE, root_dir);
+    } else if (class_name == "ZImagePipeline") {
+        m_impl = std::make_shared<ZImagePipeline>(PipelineType::TEXT_2_IMAGE, root_dir);
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/text2image_pipeline.cpp:41]: added tests

File: tests/python_tests/test_image_generation.py [hunk 11]
```diff
@@ -513,6 +515,85 @@ def callback(step, num_steps, latent):
...
+            width=512,
+            height=512,
+            num_inference_steps=2,
+            callback=callback,
+        )
+
+        assert len(callback_calls) > 0, "Callback should be called at least once"
+        assert image is not None
+
+
```
Comment (likholat) [inline tests/python_tests/test_image_generation.py:648]: please add test for `reshape()`

File: tests/python_tests/test_image_generation.py [hunk 12]
```diff
@@ -435,6 +436,7 @@ def test_text2image_blob_export_is_sdxl_only(self, image_generation_model, tmp_p
             pipe.export_model(tmp_path / "blob_model")
 
 
+@pytest.mark.xfail(reason="CVS-178687 z-image is not implemented in optimum-intel yet")
```
Comment (likholat) [inline tests/python_tests/test_image_generation.py]:
```suggestion
```
These are Flux2Klein tests

File: src/cpp/src/image_generation/zimage_pipeline.hpp [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+        m_generation_config.height = 512;
+        m_generation_config.width = 512;
+        m_generation_config.num_inference_steps = 4;
+        m_generation_config.guidance_scale = 0.0f;
+        m_generation_config.strength = m_pipeline_type == PipelineType::IMAGE_2_IMAGE ? 0.6f : 1.0f;
+    }
+
+    void check_image_size(const int height, const int width) const override {
+        OPENVINO_ASSERT(height == 512 && width == 512,
+            "ZImagePipeline only supports fixed resolution of 512x512");
```
Comment (likholat) [inline src/cpp/src/image_generation/zimage_pipeline.hpp]: I see different output shapes at the model page: https://huggingface.co/Tongyi-MAI/Z-Image-Turbo

Image

File: src/cpp/src/image_generation/zimage_pipeline.hpp [hunk 14]
```diff
@@ -0,0 +1,428 @@
...
+    }
+
+    void check_inputs(const ImageGenerationConfig& generation_config, ov::Tensor initial_image) const override {
```
Comment (likholat) [inline src/cpp/src/image_generation/zimage_pipeline.hpp:391]: We should probably call `check_image_size` here as well

File: tests/python_tests/test_image_generation.py (hunk 12 above)
Comment (pavel-esir) [reply tests/python_tests/test_image_generation.py]: applied

File: src/cpp/src/image_generation/zimage_pipeline.hpp (hunk 13 above)
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp]: Removed this redundant assert and added reshape tests

File: src/cpp/src/image_generation/zimage_pipeline.hpp (hunk 14 above)
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp:391]: done

File: tests/python_tests/conftest.py [hunk 15]
```diff
@@ -114,6 +115,7 @@ def image_generation_model(request):
     model_id = getattr(request, "param", DEFAULT_IMAGE_GEN_MODEL_ID)
     model_name = IMAGE_GEN_MODELS[model_id]
     models_dir = get_ov_cache_converted_models_dir()
+    print("MODELS DIR", models_dir)
```
Comment (pavel-esir) [inline tests/python_tests/conftest.py]: remove this

File: src/cpp/src/image_generation/zimage_pipeline.hpp [hunk 16]
```diff
@@ -0,0 +1,433 @@
...
+    }
+
+protected:
+    void initialize_generation_config(const std::string& class_name) override {
+        m_generation_config.max_sequence_length = 128;
+        m_generation_config.num_images_per_prompt = 1;
+        m_generation_config.height = 512;
+        m_generation_config.width = 512;
+        m_generation_config.num_inference_steps = 4;
+        m_generation_config.guidance_scale = 0.0f;
```
Comment (likholat) [inline src/cpp/src/image_generation/zimage_pipeline.hpp]: Shouldn't it be `5.0` by default?
https://github.com/huggingface/diffusers/blob/2f7e0154a9db246e95c9ede43edba7db5b130805/src/diffusers/pipelines/z_image/pipeline_z_image.py#L340

File: src/cpp/src/image_generation/zimage_pipeline.hpp [hunk 17]
```diff
@@ -0,0 +1,433 @@
...
+
+    ImageGenerationPerfMetrics get_performance_metrics() override {
+        return m_perf_metrics;
+    }
+
+protected:
+    void initialize_generation_config(const std::string& class_name) override {
+        m_generation_config.max_sequence_length = 128;
+        m_generation_config.num_images_per_prompt = 1;
+        m_generation_config.height = 512;
```
Comment (likholat) [inline src/cpp/src/image_generation/zimage_pipeline.hpp]: default values differ from diffusers https://github.com/huggingface/diffusers/blob/2f7e0154a9db246e95c9ede43edba7db5b130805/src/diffusers/pipelines/z_image/pipeline_z_image.py#L329

File: tests/python_tests/test_image_generation.py [hunk 18]
```diff
@@ -513,6 +527,96 @@ def callback(step, num_steps, latent):
         assert image is not None
 
 
+@pytest.mark.xfail(reason="CVS-178687 z-image is not implemented in optimum-intel yet")
```
Comment (likholat) [inline tests/python_tests/test_image_generation.py:559]: Please enable the tests before merging

File: src/cpp/src/image_generation/zimage_pipeline.hpp [hunk 19]
```diff
@@ -0,0 +1,433 @@
+// Copyright (C) 2023-2026 Intel Corporation
```
Comment (likholat) [inline src/cpp/src/image_generation/zimage_pipeline.hpp:1]: Please update docs

File: src/cpp/src/image_generation/zimage_pipeline.hpp (hunk 16 above)
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp]: `5.0` is default for `z-image` for `z-image-turbo` it's `0.0`. They both are initialized with the same python class only weights and parameters differ.

File: src/cpp/src/image_generation/zimage_pipeline.hpp (hunk 13 above)
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp]: This assert has been removed since model now works with any shape.

File: src/cpp/src/image_generation/zimage_pipeline.hpp (hunk 14 above)
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp:391]: Done

File: tests/python_tests/test_image_generation.py (hunk 12 above)
Comment (pavel-esir) [reply tests/python_tests/test_image_generation.py]: reverted back

File: tests/python_tests/test_image_generation.py (hunk 11 above)
Comment (pavel-esir) [reply tests/python_tests/test_image_generation.py:648]: Done

File: tests/python_tests/conftest.py (hunk 15 above)
Comment (pavel-esir) [reply tests/python_tests/conftest.py]: removed

File: src/cpp/src/image_generation/zimage_pipeline.hpp (hunk 17 above)
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp]: I set values so that they default to `z-image-turbo`, but since this pipeline is common for all `z-image` aligned with default values of z-image from HF.

File: src/cpp/src/image_generation/zimage_pipeline.hpp (hunk 16 above)
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp]: Answered below. Yes, defaulted to `5.0` as it's for `z-image`

File: tests/python_tests/test_image_generation.py (hunk 18 above)
Comment (pavel-esir) [reply tests/python_tests/test_image_generation.py:559]: Sure!

File: src/cpp/src/image_generation/zimage_pipeline.hpp [hunk 20]
```diff
@@ -0,0 +1,433 @@
...
+                noise_pred.copy_to(positive_noise_pred);
+                const ov::Tensor negative_noise_pred =
+                    m_transformer->step(latent, timestep, m_negative_prompt_embeds);
+
+                float* positive_noise_pred_data = positive_noise_pred.data<float>();
+                const float* negative_noise_pred_data = negative_noise_pred.data<const float>();
+                for (size_t i = 0; i < positive_noise_pred.get_size(); ++i) {
+                    positive_noise_pred_data[i] += custom_generation_config.guidance_scale *
+                                                   (positive_noise_pred_data[i] - negative_noise_pred_data[i]);
+                }
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp:321]: fixed

File: src/cpp/src/image_generation/zimage_pipeline.hpp [hunk 21]
```diff
@@ -0,0 +1,433 @@
...
+        ImageGenerationConfig custom_generation_config = m_generation_config;
+        custom_generation_config.update_generation_config(properties);
+
+        OPENVINO_ASSERT(!mask_image, "Mask image is not supported by ZImagePipeline");
+
+        const size_t vae_scale_factor = m_vae->get_vae_scale_factor();
+
+        custom_generation_config.height = custom_generation_config.height == 0 ? 512 : custom_generation_config.height;
+        custom_generation_config.width = custom_generation_config.width == 0 ? 512 : custom_generation_config.width;
+
```
[the remark this answers is not in the document]
Comment (pavel-esir) [reply src/cpp/src/image_generation/zimage_pipeline.hpp:268]: Fixed

**Review by sgonorov — APPROVED** [review]
